### PR TITLE
Bucket/Object Tags feature with versioning support included

### LINKED ISSUE #884: Question: user tags support
Hi all,

I've seen that other languages sdk ( golang, python ) support setting and getting user tags on objects. Is there any plan for it in the javascript library?

thanks

### PR DESCRIPTION
**Pull request by prakashsvmx**
Implement Tags on 
- Bucket
  - `setBucketTagging`
  - `removeBucketTagging`
  - `getBucketTagging`
- Object
  - `setObjectTagging`   ( Supports optional versionId)
  -  `removeObjectTagging` (Supports optional versionId)
  - `getObjectTagging` (Supports optional versionId)
  
Closes #884

### CODE REVIEW

File: src/main/xml-parsers.js [hunk 1]
```diff
@@ -459,4 +459,9 @@ export function parseListObjectsV2WithMetadata(xml) {
 export function parseBucketVersioningConfig(xml){
   var xmlObj = parseXml(xml)
   return xmlObj.VersioningConfiguration
+}
+
+export function parseTagging(xml){
+  const xmlObj = parseXml(xml)
+  return xmlObj
```
Comment (ebozduman) [inline src/main/xml-parsers.js:466]: `xmlObj` looks like this; `xmlObj = { Tagging: { TagSet: { Tag: [Array] } } }`

Do we really return the whole object with all these layers?

Why not return a cleaner version of it, `return xmlObj.Tagging.TagSet.Tag`?

The cleaner version would be much more similar to what is displayed in `mc tag list <alias/bucket[/path-to-object]>` command output.

**Review by ebozduman — CHANGES_REQUESTED** [review]

File: src/main/xml-parsers.js (hunk 1 above)
Comment (prakashsvmx) [reply src/main/xml-parsers.js:466]: I will update it to return (`xmlObj.Tagging.TagSet.Tag` ) an array of objects like the below: 
```
[
  { Key: 'tag2', Value: 'tag value 2' },
  { Key: 'tag3', Value: 'tag value 3' },
  { Key: 'tag4', Value: 'tag value 4' }
]
```

File: src/main/minio.js [hunk 2]
```diff
@@ -2259,6 +2259,245 @@ export class Client {
...
+
+    if(removeOpts && Object.keys(removeOpts).length && removeOpts.versionId){
+      query =`${query}&versionId=${removeOpts.versionId}`
+    }
+    const requestOptions = { method, bucketName, objectName, query }
+
+    if(objectName){
+      requestOptions['objectName']=objectName
+    }
+    this.makeRequest(requestOptions, '', 200, '', true, cb)
```
[the remark this answers is not in the document]
Comment (prakashsvmx) [reply src/main/minio.js:2385]: Thank you. I will add a comment.

**Comment by prakashsvmx** [comment]
@ebozduman thank you for the review and feedback. I have addressed the comments. PTAL.

File: src/main/minio.js [hunk 3]
```diff
@@ -2259,6 +2259,246 @@ export class Client {
...
+  /** To set Tags on a bucket or object based on the params
+     *  __Arguments__
+     * taggingParams _object_ Which contains the following properties
+     *  bucketName _string_,
+     *  objectName _string_ (Optional),
+     *  tags _object_ of the form {'<tag-key-1>':'<tag-value-1>','<tag-key-2>':'<tag-value-2>'}
+     *  putOpts _object_ (Optional) e.g {versionId:"my-object-version-id"},
+     *  cb(error)` _function_ - callback function with `err` as the error argument. `err` is null if the operation is successful.
+     */
+  setTagging(taggingParams){
```
Comment (kannappanr) [inline src/main/minio.js:2271]:
```suggestion
  setObjectTagging(taggingParams){
```

**Comment by kannappanr** [comment]
Also missing setobjecttagging example

File: src/main/minio.js (hunk 3 above)
Comment (prakashsvmx) [reply src/main/minio.js:2271]: @kannappanr  this in an internal method.  

`putObjectTagging` is the method exposed as API. 

I have followed the naming of methods similar to go sdk.

**Comment by prakashsvmx** [comment]
> Also missing setobjecttagging example
the API method to set tags on an object is: `putObjectTagging`

`putObjectTagging` example is [here](https://github.com/minio/minio-js/blob/19d83c57dbc74be158473ac2e92028803a7ce158/examples/put-object-tagging.js)

File: src/main/minio.js (hunk 3 above)
Comment (prakashsvmx) [reply src/main/minio.js:2271]: I have updated to `setObjectTagging` and the examples as well.

File: src/main/minio.js [hunk 4]
```diff
@@ -2280,6 +2280,246 @@ export class Client {
...
+
+    if(isFunction(putOpts)){
+      cb=putOpts
+      putOpts={}
+    }
+
+    if(!isObject(tags)){
+      throw new errors.InvalidArgumentError('tags should be of type "object"')
+    }
+    if(Object.keys(tags).length > 50){
```
Comment (kannappanr) [inline src/main/minio.js:2377]: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectTagging.html

The limit is 10

Comment (prakashsvmx) [reply src/main/minio.js]: I have updated. Thank you.

**Review by ebozduman — DISMISSED** [review]
LGTM

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**